### ISSUE TITLE
Refactoring ideas

### ISSUE THREAD

**Issue by sstok** [gushphp/gush#401]
Refactoring ideas
==============

This a collection of idea's for refactoring, and not completed yet :dancer: 

As some require a big changes in the application architecture there is a difference to whats actionable for the next minor release, and what will be done in Gush 2.0.

## 1.x actionable

* [x] ~~Adapt the new Symfony Command Style~~ #402
* [x] ~~Detect if cwd is a git repository (only once) (no more caching :+1: )~~ #578 
* [x] ~~Remove `ProcessHelper::runCommands()` (currently only used in GitHelper::mergeBranch() but should have proper error handling).~~  (DROPPED)
* [x] Move `ProcessHelper::probePhpCsFixer()` to its own service, and redirect output back to the console. ( #513 )
* [x] ~~The `ProcessHelper` is output-aware for logging, but instead for we can use the `DebugFormatterHelper` is especially designed for this.~~ #579 
* [x] ~~Always inform which org and repo the user is on like: "No new issues found in gushphp/gush"~~ (#522).
* [x] ~~Merge PullRequestPatOnTheBackCommand with `PullRequestMergeCommand` in confirm mode or when 'auto_thanks' is set to true.~~ ( DROPPED in favor of #572  )
* [x] ~~When configuring load all the current values, allow empty password/token when configuring, if there is already a password (ignore) require otherwise.~~ (Skipping Adapters is possible with #413 so this is not needed anymore).
* [x] Show both adapters and issue-trackers in the same list (but mention there support) {1} (#413)
  * `GitHub (Repository & Issue Tracker)`
  * `Jira (Issue Tracker)`

{1} Unless you use an Enterprise adapter its not really possible to have two separate accounts (and doesn't really make sense either :smile: )

### Enterprise support (yea!) (depends on #413)

* Add `ProvidesEnterprise` interface (better naming welcome) for AdapterFactory which adds a `createLocalConfigurator()` method, this method will create a new Configurator that ONLY handles the repository-link information (API address and repository URL) but not authentication (see snippet below for details).
* Adapter repository-link information and authentication is stored in the global gush config, name is calculated by the hashed config (`local gush.yml "adapter" value`).
* Configuring (authenticating) is done using the `core:configure` command.
   * Enterprise adapters are hidden (*see them as templates*).
   * Each Enterprise adapter can be used a multiple number of times, so each *configured* enterprise adapter is listed. Configured being, its authenticated.
   * If its not configured its only shown in the list when: you are in a working directory with a local gush file and an Enterprise adapter set as adapter (see snippet).
   * Shown in the same selection list, but displayed as (`GitHub Enterprise on [label] (RepositoryManager, IssueTracker)`)
* To *set* an Enterprise adapter run the `core:init` command, this will display all the adapters and Enterprise adapters, once you choose an Enterprise adapter it will create a new LocalConfigurator for setting the repo-url and api-url. *Because you are in configured working dir `core:configure` will automatically pick-up the configuration for authenticating :+1: *

local gush.yml file

```yaml
repo_adapter:
   type: github_enterprise
   label: My company
   config: 
     repo_url: ...
     api_url: ...
```

## 2.0

*  [x] ~~Switch to webmozart/console~~ (DROPPED)
*  [ ] Determine a Service container (Pimple, Aura.Di, etc.) (DROPPED)
*  [ ] Move all none helpers to services
*  [ ] Updating the configuration is a debug detail, we can use Console logger to make this visual when 
debugging is enabled.
*  [x] ~~Use `--org` as remote for git commands, and automatically register and fetch the remote when not present.~~
*  [x] ~~Automatically determine the org (unless set explicitly) for a forked repository; This not possible in 1.x as the Adapter must be available which is not possible at the subscriber.~~

### Configuration

*  [x] Not all adapters use a username for authentication, always ensure a username for each adapter
* ~~Configuration version-schema, when the schema is outdated force a `core:configure` (and provide a loader for to try and load older versions).~~ (DROPPED)

#### Changing commands

Commands should be threaded like Controllers, they should be slim and all extractable code should be in services (which also makes testing much easier :) )

* BranchChangelogCommand
* CoreConfigureCommand
* InitCommand (config loading/saving should be handled in a central Configuration service)
* IssueListCommand (issue filtering logic)
* MetaHeaderCommand (determine list of files to update, update the actual files)

### Adapters

*  [x] Loading the adapter happens using a Factory
*  [x] Finding the adapter to use is done using a resolver, finding the adapter in the following order:
  * local gush.yml file
  * `--adapter` option (useful when not in a Git folder)
  * Git remote (origin)
  * ~~Global gush config, default adapter.~~ (DROPPED)
* Use a value-object for referencing repositories (instead of using "org:repo")

#### Exception handling

* Adapters must throw exceptions that make sense and "not just the raw error message", only when the error is not handled we should throw an raw API exception.

#### Enterprise support (yea!)

* Adapters must be marked with enterprise support (better interface name needed).
* Adapter is only configured when its not in the global gush configuration, name is calculated by the hashed config (`local gush.yml adapter value`).
* Configuring authentication is only possible once its registered (this needs improvement, `gush config --local authentication` ?).

local gush.yml file

```yaml
repo_adapter:
   type: github_enterprise
   repo_url: ...
   api_url: ...
```

**Comment by cordoval** [comment]
he he :blush: i can add but to be actionables, we either turn every bullet here into a checkbox and also add @name here to distribute the workload

several :older_man: :older_man: can work better then a single :older_man:

**Comment by sstok** [comment]
I found a better way to improve Gush :yum: 

We already talked about https://github.com/webmozart/console as an option.
At the time there was no documentation, BUT! There is now :+1: and its looking pretty promising.

If we take webmozart/console and add a service resolver like Pimple or Aura.Di we can inject the services (not the container) directly in the CommandHandlers constructor.

Plus, by using a Service container we also add extension points for plug-ins.

As this is major change, and I also have some ideas to further improve and robust-en the adapters and error handling I think we schedule this for Gush 2.0. The list above needs some resorting to whats actionable now and what can be done in 2.0.

**Comment by sstok** [comment]
I updated the description to split 1.x actionable and 2.0.

**Comment by cordoval** [comment]
The styling has been accomplished, not sure if it was an item here or not.

**Comment by sstok** [comment]
Yes ;)

**Comment by sstok** [comment]
I improved the Enterprise support description and moved it to actionable as I have a better idea how to do this :)

**Comment by cordoval** [comment]
@stof would be proud of us about this refactor :blush:

**Comment by sstok** [comment]
I have a rather big idea (*not completed yet*) related to #462

Each adapter can be configured multiple times, the configuration itself is kept separately (transparent from the commands). 

* Rename 'core:configure' to 'configure:core' and 'core:init' to 'configure:local'
* 'configure:local' configures your local gush configuration.
* If the adapter config-indentifier ('adapter + [no authentication information, eg. repository and API URL]') is already present in the global configuration it's used, else Gush will ask for the authentication credentials.
* For the sake of usability we provide sensible defaults for the repository/API url's.
* 'core:configure' allows to list current adapter configurations, edit/remove them, and allows to add new adapter configurations.
* When you are outside a Gush configured folder Gush will require the `--[issue|repo]-adapter-config` for identifying an adapter configuration, and the `--[issue|repo]-[org|project|repo]` options.
* The git remote-name of a repository is composed as: `[canonical repository-name (eg. github.com-gushphp-gush)` and is generated by the adapter itself (to prevent including unnecessary details like `projects/` or `repositories/`). *resolves #438*

**Comment by cordoval** [comment]
we need to break up ideas into a line of smaller PRs that we can be doing, like the renaming of the commands, so we should split this into tickets like that.

**Comment by cordoval** [comment]
must be broken down into smaller tickets

**Comment by sstok** [comment]
I experimented some with Webmozard/console but atm it doesn't support short commands like `p:cr`, it's possible this can be added but I'm not sure. So I'm going to scratch that one of the list.

**Comment by cordoval** [comment]
yes i am not interested anymore in decoupling but more into
simplifying and leaving gush very slim too, so mozart is a no go

**Comment by sstok** [comment]
Removed `remove parseProcessArguments()` as there some edge cases that cannot be solved without this (mainly the EDITOR).

**Comment by sstok** [comment]
Closing as most concerns are taking care of 👍 enterprise support is still planned, and will hopefully soon be introduced :metal:

**Comment by cordoval** [comment]
👍 glad this ticket is closed, it has been a while, nice job
